### PR TITLE
Add default option

### PR DESCRIPTION
**Pull request by kyriesent**
@blake-nicholson I've merged the default value branch from https://github.com/thecsyco/node-avro-io/pull/1

There are a few tests failing, but the same tests fail in `master` so I didn't bother fixing. The new test for default values work though.

### CODE REVIEW

**Comment by kyriesent** [comment]
@blake-nicholson the `long` fix is in so those tests are passing...

But now `mocha test` gives me a few fails:

```
1) Reader() streaming should read a large avro data stream compressed with deflate:
     Uncaught ENOENT, open '/home/vagrant/src/node-avro-io/test/data/log.deflate.avro'
  Error: ENOENT, open 'test/data/log.deflate.avro'

  2) Reader() streaming should read a large avro data stream compressed with snappy:
     Uncaught ENOENT, open '/home/vagrant/src/node-avro-io/test/data/log.snappy.avro'
  Error: ENOENT, open 'test/data/log.snappy.avro
```

Looks like some test data is missing from version control. I'll open an issue for that, but can I go ahead and merge this?

**Comment by kyriesent** [comment]
Opened #20 to track the above. Otherwise, can you give me an emoji?? :smile:

**Comment by kyriesent** [comment]
Testing slack integration...

**Comment by kyriesent** [comment]
Second test of slack integration, sorry for the spam!

File: lib/io.js [hunk 1]
```diff
@@ -650,6 +650,9 @@ DatumWriter.prototype = {
                     ) {
                 schemaIndex = i;
                 break;
+            } else if (writersSchema.schemas[i].type === 'map' && writersSchema.schemas[i].validate(writersSchema.schemas[i].type, datum)) {
+                schemaIndex = i;
+                break;
```
Comment (blake-nicholson) [inline lib/io.js:655]: How is this related to the default option?  This seems like it should not be part of this PR.

File: lib/schema.js [hunk 2]
```diff
@@ -152,6 +154,16 @@ _.extend(Schema.prototype, {
...
                 }
+            } else if ( _.isArray( schema.type ) ) {
+                if (_.isEmpty(schema.type)) {
+                    throw new AvroErrors.InvalidSchemaError('unions must have at least 1 branch');
+                }
+                var schemaTypes = _.clone( schema.type );
+                var branchTypes = _.map(schemaTypes, function(type) {
+                    schema.type = type;
+                    return self.parse( schema, namespace);
+                });
```
Comment (blake-nicholson) [inline lib/schema.js:165]: Again, this block of code seems to be unrelated to enabling the default value.

File: lib/schema.js [hunk 3]
```diff
@@ -267,6 +279,10 @@ _.extend(Schema.prototype, {
                     );
                 }
                 break;
+            case 'map': 
+              if ( _.isNull( datum ) || _.isArray( datum ) || _.isFunction( datum ) || !_.isObject( datum ) ) 
+                throw new AvroErrors.DataValidationError("Data [%j] not a an object", datum, this.symbols);
+              break;
```
Comment (blake-nicholson) [inline lib/schema.js:285]: Again, not related to defaults.

File: lib/io.js (hunk 1 above)
Comment (kyriesent) [inline lib/io.js:655]: I'm not sure. This and most of the other changes were all in https://github.com/thecsyco/node-avro-io/pull/1, so I kept them. You'd probably be more qualified at this point to edit them down if you think it necessary.

File: lib/schema.js [hunk 4]
```diff
@@ -337,9 +352,10 @@ function FieldSchema(name, type) {
 
     this.name = name;
     this.type = type;
+    this.default = defaultValue;
 }
 
-//util.inherits(FieldSchema, Schema);
+util.inherits(FieldSchema, Schema);
```
Comment (blake-nicholson) [inline lib/schema.js:358]: Should FieldSchema in fact inherit from Schema?  To me, the things that inherit from Schema are valid types based upon the Avro spec.  There is no such thing as a Field type.  Why don't we just rename FieldSchema to Field and have it not inherit from Schema?

File: lib/schema.js [hunk 5]
```diff
@@ -375,14 +391,15 @@ function RecordSchema(name, namespace, fields) {
 
 util.inherits(RecordSchema, Schema);
 
-function MapSchema(type) {
+function MapSchema(type, defaultValue) {
     this.type = 'map';
     this.values = type;
+    this.default = defaultValue;
```
Comment (blake-nicholson) [inline lib/schema.js:397]: Do all schema classes have a default value?  What about storing defaultValue in the Schema class and invoking Schema.call?  I realize this doesn't save much here, but seems to be the right thing looking forward.

**Comment by kyriesent** [comment]
Closing in favor of #21
